gsl/gsl_util

```diff
@@ -135,30 +135,30 @@ inline T narrow(U u)
 // at() - Bounds-checked way of accessing static arrays, std::array, std::vector
 //
 template <class T, size_t N>
-constexpr T& at(T (&arr)[N], size_t index)
+constexpr T& at(T (&arr)[N], std::ptrdiff_t index)
 {
-    Expects(index < N);
-    return arr[index];
+    Expects(index >= 0 && index < narrow_cast<decltype(index)>(N));
+    return arr[static_cast<decltype(N)>(index)];
 }
 
 template <class T, size_t N>
-constexpr T& at(std::array<T, N>& arr, size_t index)
+constexpr T& at(std::array<T, N>& arr, std::ptrdiff_t index)
 {
-    Expects(index < N);
-    return arr[index];
+    Expects(index >= 0 && index < narrow_cast<decltype(index)>(N));
+    return arr[static_cast<decltype(N)>(index)];
 }
```

Comment:
I think it would just read easier (esp. in line 141) if it was: `static_cast<size_t>(index)` and 'narrow_cast<std::ptrdiff_t>(N)'. And to be honest, I don't think the `decltype`'s will end up saving us a lot in terms of maintenance.

Sorry to be painful, but can we just go with the boring old named types?